### PR TITLE
[com_fields] Use the color rule for the color field

### PR DESCRIPTION
**Pull request by laoneo**
### Summary of Changes
To validate the color value in the color field, the color rule should be used. The doc blocks got updated  of the `onCustomFieldsPrepareDom` function is they have been outdated.

#### Potential BC break
Additionally the color rule was changed to have the same behavior as the other rules when an empty value is passed and the field is not required. IMO this is not a BC break but a bug fixed, @Bakual, @mbabker and @wilsonge what you guys think?

### Testing Instructions
- Create a color field
- Set a wrong default value like 'nocolor'

### Expected result
An error should be shown that the field could not be saved.

### Actual result
An error should be shown that the field could not be saved.

### CODE REVIEW

File: plugins/fields/color/color.php [hunk 1]
```diff
@@ -18,4 +18,30 @@
  */
 class PlgFieldsColor extends FieldsPlugin
 {
+	/**
+	 * Transforms the field into an XML element and appends it as child on the given parent. This
```
Comment (brianteeman) [inline plugins/fields/color/color.php]: as a child

File: plugins/fields/color/color.php [hunk 2]
```diff
@@ -18,4 +18,30 @@
...
+	 * is the default implementation of a field. Form fields which do support to be transformed into
```
Comment (brianteeman) [inline plugins/fields/color/color.php]: which support being transformed into

**Comment by laoneo** [comment]
Thanks @brianteeman. I had to change the whole doc block as it was very outdated and did contain invalid information.

File: administrator/components/com_fields/libraries/fieldslistplugin.php [hunk 3]
```diff
@@ -18,9 +18,7 @@
 class FieldsListPlugin extends FieldsPlugin
 {
 	/**
-	 * Transforms the field into an XML element and appends it as child on the given parent. This
-	 * is the default implementation of a field. Form fields which do support to be transformed into
-	 * an XML Element must implement the JFormDomfieldinterface.
+	 * Transforms the field into an DOM XML element and appends it as a child on the given parent.
```
Comment (brianteeman) [inline administrator/components/com_fields/libraries/fieldslistplugin.php]: a Dom XML

File: administrator/components/com_fields/libraries/fieldsplugin.php [hunk 4]
```diff
@@ -116,9 +116,7 @@ public function onCustomFieldsPrepareField($context, $item, $field)
 	}
 
 	/**
-	 * Transforms the field into an XML element and appends it as child on the given parent. This
-	 * is the default implementation of a field. Form fields which do support to be transformed into
-	 * an XML Element must implement the JFormDomfieldinterface.
+	 * Transforms the field into an DOM XML element and appends it as a child on the given parent.
```
Comment (brianteeman) [inline administrator/components/com_fields/libraries/fieldsplugin.php]: a DOM XML

File: plugins/fields/calendar/calendar.php [hunk 5]
```diff
@@ -19,9 +19,7 @@
 class PlgFieldsCalendar extends FieldsPlugin
 {
 	/**
-	 * Transforms the field into an XML element and appends it as child on the given parent. This
-	 * is the default implementation of a field. Form fields which do support to be transformed into
-	 * an XML Element must implement the JFormDomfieldinterface.
+	 * Transforms the field into an DOM XML element and appends it as a child on the given parent.
```
Comment (brianteeman) [inline plugins/fields/calendar/calendar.php]: a DOM XML

File: plugins/fields/color/color.php [hunk 6]
```diff
@@ -18,4 +18,28 @@
  */
 class PlgFieldsColor extends FieldsPlugin
 {
+	/**
+	 * Transforms the field into an DOM XML element and appends it as a child on the given parent.
```
Comment (brianteeman) [inline plugins/fields/color/color.php]: a DOM XML

**Comment by Bakual** [comment]
> Additionally the color rule was changed to have the same behavior as the other rules when an empty value is passed and the field is not required. IMO this is not a BC break but a bug fixed, @Bakual, @mbabker and @wilsonge what you guys think?

Should be fine imho. Nothing could break anyway, it just allows an empty value now if the field is not required, which makes sense for a not-required field.

**Comment by laoneo** [comment]
If that one got merged, then I will prepare another pr which unify the required check as it is redundant code across all rules. But first let's get that one in first :wink:

**Comment by franz-wohlkoenig** [comment]
@laoneo without PR i added in color-field hex value "#000000", Article is saved with #000000

**Comment by N6REJ** [comment]
@franz-wohlkoenig thats black

**Comment by franz-wohlkoenig** [comment]
@N6REJ correct*g i referred to:

> The article saves with a wrong color value.

so i expect that #000000 got something like #123456

**Comment by N6REJ** [comment]
oh ok, thats not what you said

**Comment by laoneo** [comment]
> so i expect that #000000 got something like #123456

Why should it be transformed that way? It should keep the value.

**Comment by franz-wohlkoenig** [comment]
@laoneo thats how i understand "The article saves with a wrong color value."

**Comment by laoneo** [comment]
No it doesn't but it will not be possible to save it with invalid colors.

**Comment by franz-wohlkoenig** [comment]
@laoneo can you give a invalid hex-code to test?

**Comment by laoneo** [comment]
You can use any string as invalid hex code which doesn't represent a hex color. I'm not even sure if this can be reproduced with modifying the request as the javascript code is always reverting to a color hex code or is clearing the field when editing the article. I guess at the end of the day it has to be merged by review as the unit tests do all pass.

**Comment by franz-wohlkoenig** [comment]
so its better for Tester to adapt Test instructions to "code review"?

**Comment by laoneo** [comment]
Ok, updated the test instructions.

**Comment by laoneo** [comment]
As with the merge of #15032 it is now easier to test. Just put in a wrong default value and try to save the field.

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 892c72795e3920868021c59e71c38f00541e027a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14970">issues.joomla.org/tracker/joomla-cms/14970</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
I have tested this item :white_check_mark: successfully on 892c72795e3920868021c59e71c38f00541e027a<br /><br />An error is shown that the field could not be saved.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14970">issues.joomla.org/tracker/joomla-cms/14970</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
RTC after two successful testes.

**Comment by laoneo** [comment]
Thanks, guys!

**Pull request merged**